### PR TITLE
[MINOR][SPARKR][ML] Joint coefficients with intercept for SparkR linear SVM summary.

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?
Joint coefficients with intercept for SparkR linear SVM summary.

## How was this patch tested?
Existing tests.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala [hunk 1]
```diff
@@ -38,9 +38,17 @@ private[r] class LinearSVCWrapper private (
   private val svcModel: LinearSVCModel =
     pipeline.stages(1).asInstanceOf[LinearSVCModel]
 
-  lazy val coefficients: Array[Double] = svcModel.coefficients.toArray
+  lazy val rFeatures: Array[String] = if (svcModel.getFitIntercept) {
+    Array("(Intercept)") ++ features
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala]: In R we stack ```intercept``` with other feature names, you can refer spark.glm, spark.logit, spark.survreg.

File: R/pkg/R/mllib_classification.R [hunk 2]
```diff
@@ -124,39 +125,30 @@ setMethod("predict", signature(object = "LinearSVCModel"),
...
-            coefficients <- matrix(unlist(coefficients), ncol = nCol)
-            intercept <- callJMethod(jobj, "intercept")
+            features <- callJMethod(jobj, "rFeatures")
+            coefficients <- callJMethod(jobj, "rCoefficients")
+            coefficients <- as.matrix(unlist(coefficients))
+            colnames(coefficients) <- c("Estimate")
+            rownames(coefficients) <- unlist(features)
             numClasses <- callJMethod(jobj, "numClasses")
             numFeatures <- callJMethod(jobj, "numFeatures")
-            if (nCol == 1) {
```
Comment (yanboliang) [inline R/pkg/R/mllib_classification.R]: ML ```LinearSVC``` only supports binary classification, and will not support multiple classification in the future, so we can simplify here.

**Review by yanboliang — COMMENTED** [review]
@felixcheung I'd propose to rename ```spark.svmLinear``` to ```spark.svm```, since ```svm``` is widely used for R users by ```e1071``` package and we may support non linear model in the future (although with low probability), we can reuse this SparkR API. It would be like ```spark.gbt``` which can call two ML algorithms with the single SparkR API. What do you think of it?

File: R/pkg/R/mllib_classification.R [hunk 3]
```diff
@@ -111,10 +112,10 @@ setMethod("spark.svmLinear", signature(data = "SparkDataFrame", formula = "formu
             new("LinearSVCModel", jobj = jobj)
           })
 
-#  Predicted values based on an LinearSVCModel model
+#  Predicted values based on a linear SVM model.
```
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: I think these are intentional - we have `#  Predicted values based on an LogisticRegressionModel model`
they are prefix by `#` and not in generated doc - only for developers

File: R/pkg/R/mllib_classification.R (hunk 2 above)
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: why not label, intercept? i think they are common in R to include what goes into the model (although in many cases it just include the formula in the model summary)

**Comment by felixcheung** [comment]
are you targeting these changes for 2.2.0 -  since we are making API/return results changes here

**Comment by felixcheung** [comment]
>  I'd propose to rename spark.svmLinear to spark.svm
I see your point but `svmLinear` is also a popular name in the caret package. My concern would be coming at too general and end up having to find a strange name later because `svm` is taken, or having parameter name conflicts and so on.

Also from various threads it seems really really unlikely that we will implement non-linear form of svm like you said :)

File: R/pkg/R/mllib_classification.R (hunk 3 above)
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: there are a couple of these starting with `#`

File: R/pkg/R/mllib_classification.R (hunk 2 above)
Comment (yanboliang) [inline R/pkg/R/mllib_classification.R]: @felixcheung The change here is to make ```coefficients``` matrix has only one column named ```Estimate```. I speculate the original code referred to ```spark.logit``` which supports multiple classification, so it should have multiple columns and each columns' name should be corresponding label. However, ```LinearSVC``` will not support multiple classification in the future, so I simplified it at here.

**Comment by yanboliang** [comment]
@felixcheung Thanks for your comments. I'm targeting this for 2.2, in case for breaking change. With respect to the name issue, I'm still more prefer to rename to ```spark.svm```. There are lots of R packages which implement same functions, but we should follow the most authoritative or frequently-used packages. For SVM, I think ```e1071``` is the one we should refer, you can check the [search result of ```r svm``` in google](https://www.google.com/webhp?hl=zh-CN&ictx=2&sa=X&ved=0ahUKEwip_7ChnoPUAhWKgFQKHWKECTIQPQgD#newwindow=1&safe=strict&hl=zh-CN&q=r+svm), all items in the first page are ```e1071::svm```. To your concern about potential name conflicts, I think we can prevent it by providing parameters such as classification/regression, kernel function, loss function, etc. However, I'm still open to hear your thoughts.

**Comment by felixcheung** [comment]
@yanboliang Appreciate discussing this matter with me, and it is important to sort this out now. Normally I wouldn't mind either way; but in this case I kinda feel strongly about not making this name change for 2 main reasons:

- first, the work has been done by a contributor. I feel we are at some level undoing his work by making this change now after his work is merged, instead of providing valuable timely feedback during the review process
- second, being concise is important. I understand the popularly of the search term. Aside from future supportability, naming conflicts etc, I think we choose to name it [LinearSVC](http://people.apache.org/~pwendell/spark-nightly/spark-master-docs/spark-2.3.0-SNAPSHOT-2017_05_22_08_01-cfca011-docs/api/scala/index.html#org.apache.spark.ml.classification.LinearSVC) in Scala because it concisely describes what it does and supports. We could have named it SVM but we didn't? So I'm not sure we should name it `svm` for R. We also didn't call boosted tree `gbm` which is hugely popular, but instead `gbt`. Also, as you are aware, we get a lot of feedback and requests on adding new ML algorithm support in Spark. I think it is very important to set expectation in this case so that people does not search and find `svm` but it doesn't do what people thinks it should do? Unless you think we will go beyond linear and support polynomial etc. at some point? But I think you agree that is rather unlikely.

Anyway, what do you think?

File: R/pkg/R/mllib_classification.R [hunk 4]
```diff
@@ -46,15 +46,16 @@ setClass("MultilayerPerceptronClassificationModel", representation(jobj = "jobj"
 #' @note NaiveBayesModel since 2.0.0
 setClass("NaiveBayesModel", representation(jobj = "jobj"))
 
-#' linear SVM Model
+#' Linear SVM Model
 #'
-#' Fits an linear SVM model against a SparkDataFrame. It is a binary classifier, similar to svm in glmnet package
+#' Fits a linear SVM model against a SparkDataFrame, similar to svm in e1071 package.
+#' Currently only supports binary classification model with linear kernal.
```
Comment (bdwyer2) [inline R/pkg/R/mllib_classification.R]: Do you mean `kernel` instead of `kernal`?

**Comment by yanboliang** [comment]
@felixcheung I'm OK to keep as it is, thanks for your clarification. What about other changes in this PR?

**Review by felixcheung — APPROVED** [review]
thanks! LGTM

**Comment by felixcheung** [comment]
let's ignore the appveyor intermitted error - since it passed before simple typo changes

**Comment by yanboliang** [comment]
Merged into master and branch-2.2. Thanks for reviewing.
